Constructors cleanup in preparation of new trait encoding

adriaanm opened the pull request:
I feel it was high time to clean up a bit here. My changes are informed by the work I'm doing on the new trait encoding (fields).

Review by @retronym, cc @SethTisue & @lrytz.

adriaanm committed:
Constructors: narrow interfaces of its components

Limit communication between traits using ConstructorTransformerBase interface.
Make Constructors a bit less of a test case for itself,
by making a bunch of fields local variables of the `transformed` method.

adriaanm committed:
Constructors: consolidate `ValDef` cases in `transformed`

When we change the encoding of fields in traits,
this will come in handy...

adriaanm committed:
Constructors: mutate less

Codify the scope of mutability for various buffers,
inline one-time methods, further reduce spooky action at a distance.

adriaanm committed:
Constructors: further clarify which defs are added & when

Move conditions and template derivation back to the
call sites of `staticConstructor` (formerly `addStaticInits`).

sjrd commented on src/compiler/scala/tools/nsc/transform/Constructors.scala:
```diff
@@ -659,66 +628,106 @@ abstract class Constructors extends Statics with Transform with ast.TreeDSL {
+          case ValDef(mods, _, _, rhs) =>
+            val memoized = moveEffectToCtor(mods, rhs)
+
+            if (memoized) defBuf += deriveValDef(stat)(_ => EmptyTree)
```
This is a change of behavior, isn't it? Previously a constant type `ValDef` would have a (uninitialized and never read) field, and it doesn't have one anymore. 

I definitely approve of the change, but it would probably be good to mention this change in the commit message, and to have a test case that asserts we don't emit spurious fields anymore.

adriaanm replied:
For now, no change was intended, but it's entirely possible I made a mistake somewhere. The idea was that the old case from https://github.com/adriaanm/scala/commit/df850f0cb08742eb8af63893b028b6e367837530#diff-086eb87150c79730cf3cfe9cb8b12219L646 (`case _: ValDef if statSym.info.isInstanceOf[ConstantType] => ()`) is implemented here by returning `false` for from `moveEffectToCtor`, and thus no Valdef is emitted. You're right that the current compiler emits a field, but it doesn't seem to come from Constructors:

Both current scalac and the one from this PR do the following during Constructors for:

final.scala:

```
class C { final val C = 1 }
```

```
[[syntax trees at end of              constructors]] // final.scala
package <empty> {
  class C extends Object {
    final <stable> <accessor> def C(): Int(1) = 1;
    def <init>(): C = {
      C.super.<init>();
      ()
    }
  }
}
```

However, javap confirms your analysis:

```
Classfile /private/tmp/C.class
  Last modified Sep 1, 2015; size 552 bytes
  MD5 checksum a6c08e38b206f9e1dd72e29fc83ad9e3
  Compiled from "final.scala"
public class C
  SourceFile: "final.scala"
<snip>
{
  private final int C;
    descriptor: I
    flags: ACC_PRIVATE, ACC_FINAL
```

adriaanm replied:
In other words, I think this PR is bug compatible :-)

sjrd replied:
OK. Thanks for checking.

adriaanm replied:
I pushed a commit that explains why this happens, and included a commented-out fix for a future PR. I'm diffing bytecode of lib/reflect/comp to make sure we are indeed bytecode & bug compatible.

adriaanm committed:
Constructors: encapsulate whether to memoize a val

Triggered by sjrd's review, it turns out that we
forgot to unlink the symbol for omittable ValDefs.
I left the fix in a comment in the margin, since
this commit is meant to be bytecode-compatible.

adriaanm commented:
There is no (new) diff in bytecode between quick and strap after strap-opt (under locker.skip, and thus locker == 2.12.0-M2). Ditto for #4709, since I did this on their merge: https://github.com/adriaanm/scala/tree/namers-ctors-cleanup

lrytz commented on src/compiler/scala/tools/nsc/transform/Statics.scala:
```diff
@@ -1,49 +1,33 @@
-
     /** changes the template for the class so that it contains a static constructor with symbol fields inits,
       * augments an existing static ctor if one already existed.
       */
```
the scaladoc is outdated now

lrytz commented on src/compiler/scala/tools/nsc/transform/Constructors.scala:
```diff
@@ -318,36 +300,39 @@ abstract class Constructors extends Statics with Transform with ast.TreeDSL {
+      * TODO: XXX This condition (`isDelayedInitSubclass && remainingConstrStats.nonEmpty`) is not correct:
+      * remainingConstrStats.nonEmpty excludes too much,
+      * but excluding it includes too much.  The constructor sequence being mimicked
+      * needs to be reproduced with total fidelity.
```
this comment seems to be really far away from where the condition actually is (https://github.com/scala/scala/pull/4723/files#diff-086eb87150c79730cf3cfe9cb8b12219R707)

lrytz commented on src/compiler/scala/tools/nsc/transform/Constructors.scala:
```diff
@@ -659,66 +630,108 @@ abstract class Constructors extends Statics with Transform with ast.TreeDSL {
+        else computeOmittableAccessors(clazz, defs, auxConstructors)
+
+      // TODO: need to add the following disjunction to omittableSym to omit symbol corresponding to suppressed ValDef tree...
+      // `|| (sym.isValue && !sym.isMethod && !memoizeValue(sym))`
```
this comment i don't grok

lrytz commented:
apart from my comments about comments, looks all good

adriaanm commented:
Will incorporate your comments in my traits-late-fields branch
